File: tutorials/tour-of-restate-python/tour/part3/ticket_object.py
# Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
#
# This file is part of the Restate examples,
# which is released under the MIT license.
#
# You can find a copy of the license in the file LICENSE
# in the root directory of this repository or package or at
# https://github.com/restatedev/examples/

from restate.context import ObjectContext
from restate.object import VirtualObject


ticket = VirtualObject("TicketObject")


# <start_reserve>
@ticket.handler()
async def reserve(ctx: ObjectContext) -> bool:
    status = await ctx.get("status") or "AVAILABLE"

    if status == "AVAILABLE":
        ctx.set("status", "RESERVED")
        return True
    else:
        return False
# <end_reserve>


# <start_unreserve>
@ticket.handler()
async def unreserve(ctx: ObjectContext):
    status = await ctx.get("status") or "AVAILABLE"

    if status != "SOLD":
        ctx.clear("status")
# <end_unreserve>


# <start_mark_as_sold>
@ticket.handler("markAsSold")
async def mark_as_sold(ctx: ObjectContext):
    status = await ctx.get("status") or "AVAILABLE"

    if status == "RESERVED":
        ctx.set("status", "SOLD")
# <end_mark_as_sold>

File: java/tutorials/tour-of-restate-java/src/main/java/dev/restate/tour/part3/TicketObject.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Java,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package dev.restate.tour.part3;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.common.StateKey;
import dev.restate.tour.auxiliary.TicketStatus;

@VirtualObject
public class TicketObject {
    // <start_reserve>
    public static final StateKey<TicketStatus> STATE_KEY = StateKey.of("status", TicketStatus.class);

    @Handler
    public boolean reserve(ObjectContext ctx) {
        TicketStatus status = ctx.get(STATE_KEY).orElse(TicketStatus.Available);

        if (status.equals(TicketStatus.Available)) {
            ctx.set(STATE_KEY, TicketStatus.Reserved);
            return true;
        } else {
            return false;
        }
    }
    // <end_reserve>

    // <start_unreserve>
    @Handler
    public void unreserve(ObjectContext ctx) {
        TicketStatus status = ctx.get(STATE_KEY).orElse(TicketStatus.Available);

        if (!status.equals(TicketStatus.Sold)) {
            ctx.clear(STATE_KEY);
        }
    }
    // <end_unreserve>

    // <start_mark_as_sold>
    @Handler
    public void markAsSold(ObjectContext ctx) {
        TicketStatus status = ctx.get(STATE_KEY).orElse(TicketStatus.Available);

        if (status.equals(TicketStatus.Reserved)) {
            ctx.set(STATE_KEY, TicketStatus.Sold);
        }
    }
    // <end_mark_as_sold>

}
